python/ray/autoscaler/_private/util.py

```diff
@@ -103,38 +104,69 @@ def prepare_config(config):
     return with_defaults
 
 
-def rewrite_legacy_yaml_to_available_node_types(
-        config: Dict[str, Any]) -> Dict[str, Any]:
+def fillout_defaults(config: Dict[str, Any]) -> Dict[str, Any]:
+    defaults = _get_default_config(config["provider"])
+    defaults.update(config)
+
+    # Just for clarity:
+    merged_config = copy.deepcopy(defaults)
 
+    # Fill auth field to avoid key errors.
+    # This field is accessed when calling NodeUpdater but is not relevant to
+    # certain node providers and is thus left out of some cluster launching
+    # configs.
+    merged_config["auth"] = merged_config.get("auth", {})
+
+    # Handle merging logic if the original config was legacy-style
+    # (available_node_types unspecified).
     if "available_node_types" not in config:
-        # TODO(ameer/ekl/alex): we can also rewrite here many other fields
-        # that include initialization/setup/start commands and ImageId.
-        logger.debug("Converting legacy cluster config to multi node types.")
-        config["available_node_types"] = {
-            NODE_TYPE_LEGACY_HEAD: {
-                "node_config": config["head_node"],
-                "resources": config["head_node"].get("resources") or {},
-                "min_workers": 0,
-                "max_workers": 0,
-            },
-            NODE_TYPE_LEGACY_WORKER: {
-                "node_config": config["worker_nodes"],
-                "resources": config["worker_nodes"].get("resources") or {},
-                "min_workers": config.get("min_workers", 0),
-                "max_workers": config.get("max_workers", 0),
-            },
-        }
-        config["head_node_type"] = NODE_TYPE_LEGACY_HEAD
-        del config["min_workers"]
-    return config
+        merged_config = merge_legacy_yaml_with_defaults(merged_config)
 
+    return merged_config
 
-def fillout_defaults(config: Dict[str, Any]) -> Dict[str, Any]:
-    defaults = _get_default_config(config["provider"])
-    defaults.update(config)
-    defaults["auth"] = defaults.get("auth", {})
-    defaults = rewrite_legacy_yaml_to_available_node_types(defaults)
-    return defaults
+
+def merge_legacy_yaml_with_defaults(
+        merged_config: Dict[str, Any]) -> Dict[str, Any]:
+    """Rewrite legacy config's available node types after it has been merged
+    with defaults yaml.
+    """
+    logger.info("Converting legacy cluster config to multi node types.")
```

Comment:
```suggestion
    logger.warning("Converting legacy cluster config to multi node types.")
```
and refer the user to available_node_types examples